[5.10][Distributed] Handle dispatch thunks and superclasses in ad-hoc witness for decodeNextArguent

ktoso opened the pull request:
**Description:** We did not properly find the method we need to refer to when a distributed actor system was built using module resilience. This fixes this issue, as well as a lookup problem when the witness comes from a superclass.
**Risk:** Low
**Review by:** @xedin @slavapestov 
**Testing:** added new tests and checked in real project 
**Original PR:** https://github.com/apple/swift/pull/69242
**Radar:** rdar://116812673

ktoso committed:
[Distributed] Handle dispach thunks of decodeNextArgument in Distributed IRGen

correct way to detect dispatch thunk presence in GenDistributed

ktoso committed:
[Distributed] Correct handling super-class implemented ad-hoc requirements

ktoso committed:
handle detecting generic code better

ktoso commented:
@swift-ci please test

hborla commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -836,15 +838,35 @@ ArgumentDecoderInfo DistributedAccessor::findArgumentDecoder(
+    /// so we must use this instead of the decodeFn directly.
+   if (classDecl->hasResilientMetadata()) {
+      if (getMethodDispatch(decodeFn) == swift::MethodDispatch::Class) {
+        SILDeclRef(decodeFn).getOverriddenVTableEntry().getDecl()->getDeclContext()->dumpContext();
```
I suspect you meant to remove this call to `dumpContext()`?

ktoso replied:
Yes sorry removed now! And simplified the branches a bit -- thank you both @hborla and @xedin for review 🙏

ktoso committed:
Remove dumpContext

ktoso committed:
Simplify branching logic

ktoso commented:
@swift-ci please test

ktoso commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -693,7 +693,7 @@ void DistributedAccessor::emit() {
 
   // Add all of the substitutions to the explosion
-  if (auto *genericEnvironment = Target->getGenericEnvironment()) {
+  if (Target->isGeneric()) {
```
Heh, there is no `isGeneric()` on 5.10 but there is on main -- I'll revert this single line; functionally it is the same.

xedin replied:
Sounds good to me!

ktoso commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -693,7 +693,7 @@ void DistributedAccessor::emit() {
 
   // Add all of the substitutions to the explosion
-  if (auto *genericEnvironment = Target->getGenericEnvironment()) {
+  if (Target->isGeneric()) {
```
```suggestion
  if (auto *genericEnvironment = Target->getGenericEnvironment()) {
```

ktoso committed:
Remove change to isGeneric on 5.10

The semantics we need here are fine enough with the previous code; The `isGeneric` method does not exist on 5.10 so let's keep using what worked well before already.

ktoso commented:
@swift-ci please test
